libmeepgeom/GDSIIgeom.cpp

```diff
@@ -0,0 +1,190 @@
+%
+%  This program is free software; you can redistribute it and/or modify
+%  it under the terms of the GNU General Public License as published by
+%  the Free Software Foundation; either version 2, or (at your option)
+%  any later version.
+%
+%  This program is distributed in the hope that it will be useful,
+%  but WITHOUT ANY WARRANTY; without even the implied warranty of
+%  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+%  GNU General Public License for more details.  %
+%  You should have received a copy of the GNU General Public License
+%  along with this program; if not, write to the Free Software Foundation,
+%  Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+*/
+
+/***************************************************************/
+/* GDSII.cpp  -- libmeepgeom code to support meep geometry     */
+/*            -- definitions from GDSII files                  */
+/* homer reid -- 5/2018                                        */
+/***************************************************************/
+
+#include <vector>
+#include <string>
+#include "meepgeom.hpp"
+
+#ifdef HAVE_CONFIG_H
+  #include "config.h"
+#endif
+
+#ifdef HAVE_LIBGDSII
+  #include <libGDSII.h>
+#endif
+
+namespace meep_geom {
+
+#ifdef HAVE_LIBGDSII
+
+void get_polygon_bounding_box(dVec vertex_coordinates, meep::vec &max_corner, meep::vec &min_corner)
+{
+  double xmax=vertex_coordinates[2*0+0], xmin=xmax;
+  double ymax=vertex_coordinates[2*0+1], ymin=ymax;
+  for(size_t nv=0; nv<vertex_coordinates.size()/2; nv++)
+   { double x = vertex_coordinates[2*nv+0], y=vertex_coordinates[2*nv+1];
+     xmax=fmax(xmax,x); ymax=fmax(ymax,y);
+     xmin=fmin(xmin,x); ymin=fmin(ymin,y);
+   }
+  max_corner.set_direction(meep::X, xmax);
+  max_corner.set_direction(meep::Y, ymax);
+  min_corner.set_direction(meep::X, xmin);
+  min_corner.set_direction(meep::Y, ymin);
+}
+
+void get_polygon_center_size(dVec vertex_coordinates, meep::vec &center, meep::vec &size)
+{
+  meep::vec max_corner, min_corner;
+  get_polygon_bounding_box(vertex_coordinates, max_corner, min_corner);
+
+  center.set_direction(meep::X, 0.5*(max_corner.in_direction(meep::X) + min_corner.in_direction(meep::X)));
+  center.set_direction(meep::Y, 0.5*(max_corner.in_direction(meep::Y) + min_corner.in_direction(meep::Y)));
+  center.set_direction(meep::Z, 0.0);
+
+  size.set_direction(meep::X, max_corner.in_direction(meep::X) - min_corner.in_direction(meep::X));
+  size.set_direction(meep::Y, max_corner.in_direction(meep::Y) - min_corner.in_direction(meep::Y));
+  size.set_direction(meep::Z, 0.0);
+}
+
+// Search the geometry for a polygon on a given layer containing 
+// (the reference point of) a given text label.
+// If Text==NULL, find any polygon on the given layer.
+// If Layer==-1, search all layers.
+// If multiple matching polygons are found, choose one arbitrarily.
+dVec get_polygon(char *GDSIIFile, char *Text, int Layer=-1)
```

Comment:
`char*` arguments should be `const char*` if the string isn't modified (otherwise C++ doesn't like you passing literal strings these days)